### ISSUE TITLE
Details of algorithm

### ISSUE THREAD

**Issue by allefeld** [jtalbot/Labeling#1]
I'm trying to implement your complete 2010 algorithm in Python (there is an implementation, but it does not include *legibility*).

Reading the paper, there are two points I'm stuck on:

1) What are the exact rules for number formatting according to the different formats in Table 3?

2) What is the recommended value for the target density rho_t? The paper only mentions that it should be adapted for different displays.

Regarding the first point, I looked at your implementations in R (which doesn't include *legibility* either) and in C# here. I suppose that the relevant code is in [Format.cs](https://github.com/jtalbot/Labeling/blob/master/Layout/Formatters/Format.cs), but my (nonexistent) understanding of C# didn't allow me to distinguish the 8 different format types in the code, let alone reproduce them. Would it be possible to provide an explanation in natural language, or in terms of traditional C format specifiers?

**Comment by jtalbot** [comment]
It's been a number of years since I looked at this, so we may have to reconstruct the behavior together.

The legibility factor is primarily driven by whether the strings abut or overlap on the screen, which usually only happens if you choose the density of labels to be too high or the plot is very small. I didn't implement this factor in R because I couldn't figure out how to check if the actual pixel extents of plotted strings overlapped in R's standard graphics library. You may want to check that you can do that in your target environment before going to the work to add this.

If strings do abut or overlap, the format factor of the legibility score tries a variety of formats to shrink the strings to increase the space between them. There's nothing particularly special about the formats themselves. I just tried a few and assigned them scores based on my own subjective judgment about how readable they were.

The "Decimal" format is something like "%f" in C. However, note that the labels generated by section 3.2 of the paper are designed to have short decimal forms. So, we might get labels like "0.125", but never labels like "0.33333333333333...". To get the benefit of this, you will probably want to generate and store the labels in a data type that isn't subject to usual floating point imprecision. In the C# implementation I used their decimal data type, but you could also store the labels as scaled integers with an associated number of decimal places.

The factored form pulls out the string "thousand", "million", "tenths", "hundredths", etc. and puts it in the axis title. So you would get something like:
0, 5, 10, 15, 20
Price (thousands)

The K and M forms put "K" or "M" after every label:
0K, 5K, 10K, 15K, 20K

While the factored K and M forms put the "K" or "M" after the axis title:
0, 5, 10, 15, 20
Price (K)

For the scientific formats I used a common power of ten in both the factored and unfactored forms. A changing exponent is a key aspect of log axis labeling and I didn't want viewers to mistake a linear axis for a log one. For each label I found the power of ten smaller than that label (50 => 10), then I averaged the power (exponent) across all the labels and then rounded that to the nearest integer exponent and used that as the common one. There're probably much more clever ways to do this.

The 0-extend factor scores labels with matching decimal places, e.g. [0.0, 0.5, 1.0] over truncated labels, [0, 0.5, 1] since I found the former easier to read, particularly on vertical axes.

The setting of rho_t very much depends on the visual style of the plot. Some people prefer dense labels, others sparse. If you have a very large font size (e.g. for a Powerpoint slide), then you need a lot more space between labels than on a printed page. For me, at a typical 8-10pt font on a computer screen, I would say about 1 label/inch is a reasonable starting point. Of course, there is always a fairly loose relationship between the units your graphics system allows you to layout in and the physical distances a viewer actually sees after rendering, so you try to shoot for something that works on the average display.

What else can I clarify?

**Comment by allefeld** [comment]
Thanks, that's exactly the detail I was hoping for! And sorry for the double message; I suspected you didn't read messages here anymore.

The recommendation to use a decimal data type was spot on. I don't use it for the algorithm itself, but I compute the final tick values from the integer indices and exact q-values as Python `Decimal`s, and generating "Decimal" format labels from those is straightforward.

The target environment for rendering is LaTeX/TikZ, and getting the size of labels out of LaTeX is possible, but quite complicated. I think I will fix the font and then extract metrics for those characters that can appear in labels, so I can compute label extents myself.

I'll proceed implementing the formats; it's possible I'll get back to you with more questions.

**Comment by allefeld** [comment]
Another question: To me there seems to be an inconsistency between `coverage` and `coverage.max` in the R code.

If `span > range`, then `coverage.max` does not return the actual maximal value of `coverage`, which is always `1 - 0.5 * (half^2 + half^2) / ((0.1 * range)^2)`, but 1. Now this could be explained by saying that in that case, coverage is perfect. However, the same would hold for the `coverage` function itself. In any case, `coverage.max` does not seem to provide the exact pruning of a search branch w.r.t. `coverage`.

````
.coverage <- function(dmin, dmax, lmin, lmax)
{
	range <- dmax-dmin
	1 - 0.5 * ((dmax-lmax)^2+(dmin-lmin)^2) / ((0.1*range)^2)
}

.coverage.max <- function(dmin, dmax, span)
{
	range <- dmax-dmin
	if(span > range)
	{
		half <- (span-range)/2
		1 - 0.5 * (half^2 + half^2) / ((0.1 * range)^2)
	}
	else
	{
		1
	}
}
````

**Comment by allefeld** [comment]
And another question, regarding `density`: I'm not clear why the denominator is different in the formulas for `r` and `rt`.
````
.density <- function(k, m, dmin, dmax, lmin, lmax)
{
	r <- (k-1) / (lmax-lmin)
	rt <- (m-1) / (max(lmax,dmax)-min(dmin,lmin))
	2 - max( r/rt, rt/r )
}
````
The second seems to take into account the possibility that the labels are *not loose*, i.e. the displayed range goes from `dmin` to `dmax` if it is larger than the one from `lmin` to `lmax`. But if the user specifies `m`, they will probably do so on the basis of `dmin` to `dmax`, so to me it seems that the second formula should simply be `rt <- (m-1) / (dmax-dmin)`.

***

I tend towards the idea to replace `m` by a directly specified `rt` anyway, which would then be in ticks / inch. In that case `r` would have to be computed with the same dimension. Assuming that numerically the axis goes from `min(dmin,lmin)` to `max(lmax,dmax)` and has a physical length `length` in inches, the range spanned by the ticks would be
````
length * (lmax-lmin) / (max(lmax,dmax)-min(dmin,lmin))
````
and the actual density should therefore be computed as
````
r <- (k-1) / (length * (lmax-lmin)) * (max(lmax,dmax)-min(dmin,lmin))
````
– is that correct?

***

I think I figured it out. Comparing
````
r <- (k-1) / (length * (lmax-lmin)) * (max(lmax,dmax)-min(dmin,lmin))
````
to a specified `rt` is actually equivalent to your implementation if
````
m <- rt * length + 1
````
except `m` is no longer constrained to be an integer.

**Comment by allefeld** [comment]
I think I figured the last one out. Using
````
r <- (k-1) / (length * (lmax-lmin)) * (max(lmax,dmax)-min(dmin,lmin))
````
compared to a specified `rt` is actually equivalent to your implementation if
````
m <- rt * length + 1
````
except `m` is no longer constrained to be an integer.

**Comment by jtalbot** [comment]
On `.coverage`, are you asking about the `span > range` case or the `span <= range` case? The latter is the case that returns 1. To be correct, the `.coverage.max` functions only need to return an upper bound, so I think returning 1 is always safe here. But finding a tighter upper bound could make the optimization faster.

Looking at this with fresh eyes, I wonder if the following might work:

```
.coverage.max <- function(dmin, dmax, span)
{
	range <- dmax-dmin
	half <- abs(span-range)/2
	1 - 0.5 * (half^2 + half^2) / ((0.1 * range)^2)
}
```

The idea behind using `half` here is that the coverage score is maximized if the extreme ticks are equi-distant from the extreme data points. I _think_ that holds true if the tick marks are inside the data range, as well as outside.

On `.density`, your approach sounds right. I vaguely remember adding the min and max to `rt`, but I can't remember why I did it. When I implemented the R version, I had to convert to a 'number of ticks' parameterization so I could compare directly to the other algorithms; I think specifying `rt` directly is a better idea.

**Comment by allefeld** [comment]
> I _think_ that holds true if the tick marks are inside the data range, as well as outside.

I agree, and I implemented it now this way; it seems to work well.

**Comment by allefeld** [comment]
I finished my implementation. In case you are interested, it is documented [here](https://allefeld.github.io/pytikz/tikz/extended_wilkinson.html).

I allowed myself a few limitations, in particular I only implemented the 'Decimal' and 'Factored scientific' formats; I don't think a common factor should be in the labels, and I don't really like '(thousands)', '(K)', and '(M)'. It shouldn't be hard to add these formats though. As far as I know, apart from your C# code this is still the most comprehensive implementation.

The code *using* the algorithm isn't finished yet, but first results look promising:
<img width="428" src="https://user-images.githubusercontent.com/1738353/88067591-19043e80-cb67-11ea-94b0-23d14944b53b.png">

Again, thanks a lot for your help! It is uncommon that an author provides such quick and comprehensive information.
